### PR TITLE
Avoid jar-name collisions when building uberwar

### PR DESCRIPTION
**Pull request by dpiliouras**
This mainly addresses #216, but also removes reflection.

### CODE REVIEW

**Comment by weavejester** [comment]
Maybe putting this PR on another branch would be a good idea, so the PR changes don't conflict with your custom deployment?

**Comment by dpiliouras** [comment]
I had to release another version of `dpiliouras/lein-ring` yesterday, in order to fix a `ClassCastException` when running `lein ring server SOME_PORT`. At this point, I don't see why I would want to release any other versions, as from a functionality perspective everything works as expected (i.e. I've not broken anything, and the jar naming collision is gone). Therefore, I would expect that any changes we make as part of the PR process, wouldn't really be detectable by consumers (e.g. I should be able to swap my custom deployment with the official one, even if the code is slightly different).

If you think that you have a fundamentally different approach to solving the jar naming collision, I'm more than happy to discuss and/or compare it.

**Comment by weavejester** [comment]
Do you want me to go ahead and review the PR then?

**Comment by dpiliouras** [comment]
Yes please :)

File: project.clj [hunk 1]
```diff
@@ -1,12 +1,15 @@
-(defproject lein-ring "0.12.5"
+(defproject dpiliouras/lein-ring "0.12.5"
```
Comment (weavejester) [inline project.clj:1]: Project name needs to be changed back

File: project.clj [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
   :description "Leiningen Ring plugin"
-  :url "https://github.com/weavejester/lein-ring"
+  :url "https://github.com/dpiliouras/lein-ring"
```
Comment (weavejester) [inline project.clj:3]: Likewise the URL

File: project.clj [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
   :scm {:name "git"
-        :url "https://github.com/weavejester/lein-ring"}
+        :url "https://github.com/dpiliouras/lein-ring"}
   :license {:name "Eclipse Public License"
             :url "http://www.eclipse.org/legal/epl-v10.html"}
   :dependencies [[org.clojure/core.unify "0.5.7"]
                  [org.clojure/data.xml "0.0.8"]
                  [leinjacker "0.4.2"
                   :exclusions [org.clojure/core.unify]]]
+  :deploy-repositories [["clojars" {:url "https://clojars.org/repo"
```
Comment (weavejester) [inline project.clj:12]: This can be removed. Pretty much all changes to the `project.clj` file can be reverted.

File: src/leiningen/ring/server.clj [hunk 4]
```diff
@@ -64,7 +64,7 @@
                           :handler ~handler)]
        (doseq [port-file# ["target/repl-port" ".nrepl-port"]]
          (-> port-file#
-             java.io.File.
+             io/file
```
Comment (weavejester) [inline src/leiningen/ring/server.clj:67]: What's the purpose of this change?

File: src/leiningen/ring/server_headless.clj [hunk 5]
```diff
@@ -6,4 +6,4 @@
   ([project]
      (server-task project {:open-browser? false}))
   ([project port]
-     (server-task project {:port (Integer. port), :open-browser? false})))
+     (server-task project {:port (int port), :open-browser? false})))
```
Comment (weavejester) [inline src/leiningen/ring/server_headless.clj:9]: What's the purpose of this change?

File: src/leiningen/ring/uberwar.clj [hunk 6]
```diff
@@ -1,7 +1,10 @@
 (ns leiningen.ring.uberwar
   (:require [leiningen.ring.war :as war]
-            [clojure.java.io :as io])
-  (:import [java.util.jar JarFile JarEntry]))
+            [clojure.java.io :as io]
+            [clojure.string :as str])
+  (:import [java.util.jar JarFile JarEntry]
+           (java.util Properties)
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:6]: Mismatched brackets - should be `[]`.

File: src/leiningen/ring/uberwar.clj [hunk 7]
```diff
@@ -11,27 +14,45 @@
...
   (if-let [get-cp (resolve 'leiningen.core.classpath/get-classpath)]
     (get-cp project)
-    (->> (:library-path project) io/file .listFiles (map str))))
-
-(defn contains-entry? [^java.io.File file ^String entry]
-  (with-open [jar-file (JarFile. file)]
-    (some (partial = entry)
-          (map #(.getName ^JarEntry %)
-               (enumeration-seq (.entries jar-file))))))
+    (->> (:library-path project) (io/file) .listFiles (map str))))
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:17]: Change `(io/file)` to `io/file` too be consistent.

File: src/leiningen/ring/uberwar.clj [hunk 8]
```diff
@@ -11,27 +14,45 @@
...
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet?  #{"javax/servlet/Servlet.class"})
+(def ^:private pom-properties? #(str/ends-with? % "pom.properties"))
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:25]: Use `defn-` for private functions.

File: src/leiningen/ring/uberwar.clj [hunk 9]
```diff
@@ -11,27 +14,45 @@
...
+
+(defn- war-path-for-jar
+  [^java.io.File jar]
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:28]: Arguments should fit on the same line as the function name where there's room to do so.

File: src/leiningen/ring/uberwar.clj [hunk 10]
```diff
@@ -11,27 +14,45 @@
...
+  (with-open [jar-file (JarFile. jar)]
+    (let [javax&pom (->> (.entries jar-file)
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:30]: This function could probably be split up into two or three smaller ones.

File: src/leiningen/ring/uberwar.clj [hunk 11]
```diff
@@ -11,27 +14,45 @@
...
+              {:strs [groupId artifactId version]} (when pom-properties-full-path
+                                                     (with-open [in (.openStream (URL. pom-properties-full-path))]
+                                                       (->> (doto (Properties.) (.load in))
+                                                            (into {}))))]
+          (->> (if (and groupId artifactId version)
+                 [groupId \- artifactId \- version ".jar"] ;; found pom.properties - reconstruct the jar name to avoid collisions (issue #216)
+                 [(.getName jar)])                         ;; fallback to using the original jar name (best-effort approach)
+               (apply str "WEB-INF/lib/")))))))
+
+(defn jar-entries
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:49]: Is this intended to be part of the public API?

File: src/leiningen/ring/uberwar.clj [hunk 12]
```diff
@@ -43,3 +64,12 @@
        project war-name
        :profiles-to-merge [:uberjar]
        :additional-writes jar-entries)))
+
+(comment
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:68]: Comment should be removed.

File: src/leiningen/ring/util.clj [hunk 13]
```diff
@@ -31,7 +31,8 @@
           "  :ring {:handler your.app/handler}"))
     (lein-core/exit 1)))
 
-(defn source-file [project namespace]
+(defn source-file
+  ^java.io.File [project namespace]
```
Comment (weavejester) [inline src/leiningen/ring/util.clj:35]: No need to split this onto a new line.

File: src/leiningen/ring/war.clj [hunk 14]
```diff
@@ -169,14 +169,15 @@
                       `(~(generate-resolve destroy-sym)))))))
       :print-meta true)))
 
-(defn create-war [project file-path]
+(defn create-war
+  ^JarOutputStream [project ^String file-path]
```
Comment (weavejester) [inline src/leiningen/ring/war.clj:173]: No need to split this onto its own line.

File: src/leiningen/ring/war.clj [hunk 15]
```diff
@@ -23,7 +23,7 @@
     (.mkdirs (io/file target-dir))
     (str target-dir "/" war-name)))
 
-(defn skip-file? [project war-path file]
+(defn skip-file? [project war-path ^java.io.File file]
```
Comment (weavejester) [inline src/leiningen/ring/war.clj:26]: If `java.io.File` is going to be used for many type hints, add it to the imports and use `^File`.

File: src/leiningen/ring/war/manifest.clj [hunk 16]
```diff
@@ -35,7 +35,8 @@
         (not (coll? (second e1)))))
     (seq mf)))
 
-(defn make-manifest [project]
+(defn make-manifest
```
Comment (weavejester) [inline src/leiningen/ring/war/manifest.clj:38]: No need to put this on its own line.

**Review by weavejester — CHANGES_REQUESTED** [review]

File: project.clj (hunk 3 above)
Comment (dpiliouras) [reply project.clj:12]: My bad, i will revert all the changes to `project.clj` asap.

File: src/leiningen/ring/server.clj (hunk 4 above)
Comment (dpiliouras) [reply src/leiningen/ring/server.clj:67]: avoiding reflection

File: src/leiningen/ring/server_headless.clj (hunk 5 above)
Comment (dpiliouras) [reply src/leiningen/ring/server_headless.clj:9]: this is wrong - i've already reverted this in server.clj - will revert here too (just needs a type-hint)

File: src/leiningen/ring/war.clj (hunk 15 above)
Comment (dpiliouras) [reply src/leiningen/ring/war.clj:26]: agreed

File: src/leiningen/ring/war.clj (hunk 14 above)
Comment (dpiliouras) [reply src/leiningen/ring/war.clj:173]: ok

File: src/leiningen/ring/util.clj (hunk 13 above)
Comment (dpiliouras) [reply src/leiningen/ring/util.clj:35]: ok

File: src/leiningen/ring/uberwar.clj (hunk 12 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:68]: Not sure what the harm is, but ok

File: src/leiningen/ring/uberwar.clj (hunk 11 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:49]: no, it can be private

File: src/leiningen/ring/uberwar.clj (hunk 12 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:74]: are you sure you want this comment removed? It's a quick and easy way to test the fix.

**Comment by dpiliouras** [comment]
I believe that's all your feedback addressed (minus that little `comment` expression)

File: src/leiningen/ring/uberwar.clj (hunk 12 above)
Comment (weavejester) [reply src/leiningen/ring/uberwar.clj:80]: Can you explain the purpose of the comment? It's not obvious to me why it was left in.

Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:80]: This project has zero tests. I have added a new function which changes the jar-name as it's being copied into the uberwar. I was using the expression inside the comment as a way to test my work during development, and left it in so that you can test/try it yourself (or anyone else that might come back to it in the future).  That is all...why would you not want it in? you've not asked me to remove any other comments, so why is this special?

**Comment by dpiliouras** [comment]
are we ok here?

File: src/leiningen/ring/uberwar.clj [hunk 17]
```diff
@@ -13,25 +17,54 @@
...
-        :when (and (.endsWith fname ".jar")
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet? #{"javax/servlet/Servlet.class"})
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:25]: Rather than a function, this would be better as a constant. e.g. `servlet-class-resource`.

File: src/leiningen/ring/uberwar.clj [hunk 18]
```diff
@@ -13,25 +17,54 @@
...
+(defn- pom-properties? [file-name]
+  (str/ends-with? file-name "pom.properties"))
+
+(defn- read-jar-pom-properties [^String full-path]
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- from-pom-properties
+  "Returns a jar name which includes the groupId, artifactId, & version,
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:36]: Docstrings of private functions should be omitted. Docstrings are for public APIs. If the purpose of a function isn't apparent from its name and content, it's an indication it should be refactored.

File: src/leiningen/ring/uberwar.clj [hunk 19]
```diff
@@ -13,25 +17,54 @@
...
+
+(defn- pom-properties? [file-name]
+  (str/ends-with? file-name "pom.properties"))
+
+(defn- read-jar-pom-properties [^String full-path]
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- from-pom-properties
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:35]: A better name would be `jar-name-from-pom-properties`.

File: src/leiningen/ring/uberwar.clj [hunk 20]
```diff
@@ -13,25 +17,54 @@
...
+  "Returns a jar name which includes the groupId, artifactId, & version,
+   as read by the pom.properties file inside the jar file found at <jar-path>."
+  [jar-path pom-properties-jar-path]
+  (let [pom-properties-full-path (str "jar:file:" jar-path "!/" pom-properties-jar-path)
+        {:strs [groupId artifactId version]} (read-jar-pom-properties pom-properties-full-path)]
+    (when (and groupId artifactId version)
+      ;; found pom.properties - reconstruct the jar name to avoid collisions (issue #216)
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:42]: I think we can remove this comment as the associated commit will explain why.

File: src/leiningen/ring/uberwar.clj [hunk 21]
```diff
@@ -13,25 +17,54 @@
...
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- from-pom-properties
+  "Returns a jar name which includes the groupId, artifactId, & version,
+   as read by the pom.properties file inside the jar file found at <jar-path>."
+  [jar-path pom-properties-jar-path]
+  (let [pom-properties-full-path (str "jar:file:" jar-path "!/" pom-properties-jar-path)
+        {:strs [groupId artifactId version]} (read-jar-pom-properties pom-properties-full-path)]
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:40]: Lines should be within 80 characters in length.

File: src/leiningen/ring/uberwar.clj [hunk 22]
```diff
@@ -13,25 +17,54 @@
...
+    (when (and groupId artifactId version)
+      ;; found pom.properties - reconstruct the jar name to avoid collisions (issue #216)
+      (str groupId \- artifactId \- version ".jar"))))
+
+(defn- war-path-for-jar [^File jar]
+  (with-open [jar-file (JarFile. jar)]
+    (let [javax&pom (->> (.entries jar-file)
+                         enumeration-seq
+                         (map (fn [^JarEntry je] (.getName je)))
+                         (filter (some-fn javax-servlet? pom-properties?)))] ;; 2 elements maximum
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:50]: Why 2 elements maximum? What is this comment conveying? This threading could also be split into a function that would explain what it does.

File: src/leiningen/ring/uberwar.clj [hunk 23]
```diff
@@ -13,25 +17,54 @@
...
+      ;; Servlet container will have it's own servlet-api impl
+      (when-not (some javax-servlet? javax&pom)
+        ;; we've confirmed that one of the two possible files is missing,
+        ;; so if we find one (via `first`), it will be the pom.properties
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:54]: If we need a comment explaining reasoning, it should be all in one place, explaining what it's doing in abstract terms. Comments ideally shouldn't be used to explain what an individual line of code is doing.

File: src/leiningen/ring/uberwar.clj [hunk 24]
```diff
@@ -13,25 +17,54 @@
...
+        (str "WEB-INF/lib/"
+             (or (some->> (first javax&pom)
+                          (from-pom-properties (.getAbsolutePath jar)))
+                 ;; fallback to using the original jar name (best-effort approach)
+                 (.getName jar)))))))
+
+(defn- jar-entries
+  "Populates the 'WEB-INF/lib/' directory of the given <war> with this project's dependencies (jars).
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:62]: This docstring should be removed, and a better name for the function should be given. Perhaps something like `populate-war-with-dependent-jars`.

**Review by weavejester — CHANGES_REQUESTED** [review]
I've added some more feedback. There's also a couple of comments that weren't addressed previously that I've left in. The code is looking better, but we're not quite there.

File: src/leiningen/ring/uberwar.clj (hunk 17 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:25]: Sorry, I'm not quite following this...do you mean change the name, or the actual value (i.e. the 1-element set)?

File: src/leiningen/ring/uberwar.clj (hunk 22 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:50]: Do you want me to remove the comment or expand on it?

File: src/leiningen/ring/uberwar.clj (hunk 17 above)
Comment (weavejester) [reply src/leiningen/ring/uberwar.clj:25]: Something like:

```clojure
(def ^:private `servlet-class-resource` "javax/servlet/Servlet.class")
```

The reason is that `#{servlet-class-resource}` then gives more information to the reader. They know that you're matching against an exact string, whereas `javax-servlet?` is opaque - they'd need to read the definition to understand what it does.

File: src/leiningen/ring/uberwar.clj (hunk 22 above)
Comment (weavejester) [reply src/leiningen/ring/uberwar.clj:50]: I don't understand why you're searching for both the servlet class and the pom.properties file at the same time. Why not just compile a list of entries, then do a search for one, then a search for the other?

```clojure
(defn- war-path-for-jar [^File jar]
  (with-open [jar-file (JarFile. jar)]
    (let [entries (find-jar-entry-names jar-file)]
      (when-not (some #{javax-servlet-class} entries)
        (path-from-pom-properties (find-pom-properties entries))))))
```

Or more abstractly:

```clojure
(defn- war-path-for-jar [^File jar]
  (with-open [jar-file (JarFile. jar)]
    (when-not (contains-javax-servlet-class? jar-file)
      (path-from-pom-properties (read-pom-properties jar-file)))))
```

File: src/leiningen/ring/uberwar.clj (hunk 17 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:25]: ok, I will do this, but bear in mind that I do want to use this as a predicate (twice), so one or another (i.e. here or below), i will have to def that set, so it's not clear to me what the benefit is.  In any case, I'll follow your suggestion...

File: src/leiningen/ring/uberwar.clj (hunk 22 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:50]: Searching for both in one-pass, is an optimisation, simply because JARs can (and often do) contain thousands of files. That is partly why the `uberwar` task takes so long. Repeating the search is quite wasteful i think. If you insist, I will split it, but please consider the implications. This search (and compiling) make up the majority of the time spent.

File: src/leiningen/ring/uberwar.clj [hunk 25]
```diff
@@ -13,25 +17,49 @@
...
+                          (jar-name-from-pom-properties (.getAbsolutePath jar)))
+                 ;; fallback to using the original jar name (best-effort approach)
+                 (.getName jar)))))))
+
+(defn- populate-war-with-dependent-jars
+  [war project]
   (doseq [jar-file (jar-dependencies project)]
-    (let [dir-path (.getParent jar-file)
-          war-path (war/in-war-path "WEB-INF/lib/" dir-path jar-file)]
+    (when-let [war-path (war-path-for-jar jar-file)]
```
[the remark this answers is not in the document]
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:62]: Excluding dependencies that contain the javax.servlet class was already the case (see `jar-dependencies`), before I ever laid hands on this project. The `;; Servlet container will have it's own servlet-api impl` comment is not mine - if you look for it in the existing codebase, you'll see that right under it there is a `(not (contains-entry? ...` expression.  So the answer to your question is no - I'm doing that because I don't want to affect what jars are included, only their names.

File: src/leiningen/ring/uberwar.clj (hunk 22 above)
Comment (weavejester) [reply src/leiningen/ring/uberwar.clj:50]: In general, we should optimise first for readability and then for performance, if necessary. Readability is often more important, but it is also easier to evaluate at a glance. Performance can be counter-intuitive, and generally you want to be running benchmarks as you go to to check your assumptions.

Since we're looking for an exact match in the case of looking for the servlet class, my first thought was to check to see if the `JarFile` class has an index lookup and then benchmark that. I ran some tests with a jruby jar I had lying around, as that has 13,000 entries:

```clojure
user=> (quick-bench (.getEntry j "org/jruby/util/io/EncodingUtils.class"))
Execution time mean : 766.012522 ns

user=> (quick-bench (some #{"org/jruby/util/io/EncodingUtils.class"} (map (fn [^java.util.jar.JarEntry je] (.getName je)) (enumeration-seq (.entries j)))))
Execution time mean : 3.253957 ms
```

You can see `.getEntry` is around 4000x faster. So if we want to optimize checks for the servlet class, then we'd write a function like:

```clojure
(defn- contains-javax-servlet-class? [^JarFile jar-file]
  (some? (.getEntry jar-file "javax/servlet/Servlet.class")))
```

Performance should start with careful reading of the classes and methods being used, and then benchmarked to confirm our assumptions.

Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:50]: I won't argue with anything you said, but I will say that  this is how it should have been implemented in the first place. I was only trying to make the least amount of (potentially breaking) changes. I wasn't really trying to make things faster in the code I added - I was merely trying to **not** make them slower. I will refactor this first thing tomorrow morning - thanks 👍

Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj]: This is done

**Comment by dpiliouras** [comment]
All the code snippets under which you commented now show as `Outdated`, so I think I've addressed everything. Have a great weekend...

**Comment by dpiliouras** [comment]
Are we OK here? No pressure - just making sure we're not deadlocked (i.e. waiting on each other)...

**Comment by weavejester** [comment]
Ah, my apologies - I thought I had reviewed it. Let me go through it now...

File: src/leiningen/ring/uberwar.clj [hunk 26]
```diff
@@ -1,7 +1,11 @@
 (ns leiningen.ring.uberwar
   (:require [leiningen.ring.war :as war]
-            [clojure.java.io :as io])
-  (:import [java.util.jar JarFile JarEntry]))
+            [clojure.java.io :as io]
+            [clojure.string :as str])
+  (:import [java.util.jar JarFile JarEntry]
+           [java.util Properties]
+           [java.net URL]
+           (java.io File)))
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:8]: This `(java.io File)` should be `[java.io File]` for consistency.

File: src/leiningen/ring/uberwar.clj [hunk 27]
```diff
@@ -13,26 +17,48 @@
...
-    (let [dir-path (.getParent jar-file)
-          war-path (war/in-war-path "WEB-INF/lib/" dir-path jar-file)]
-      (war/file-entry war project war-path jar-file))))
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(defn- contains-javax-servlet-class? [^JarFile jar-file]
+  (some? (.getEntry jar-file "javax/servlet/Servlet.class")))
+
+(defn- pom-properties? [^JarEntry entry]
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:28]: This is not a predicate, so the name shouldn't end with a `?`. It's probably better to split this into two functions:

```clojure
(defn- pom-properties? [^JarEntry entry]
  (str/ends-with? (.getName entry) "pom.properties"))

(defn- find-pom-properties [^JarFile jar-file]
  (->> jar-file .entries enumeration-seq (filter pom-properties?) first))
```

File: src/leiningen/ring/uberwar.clj [hunk 28]
```diff
@@ -13,26 +17,48 @@
...
+  (when (-> (.getName entry)
+            (str/ends-with? "pom.properties"))
+    entry))
+
+(defn- read-jar-pom-properties
+  [^JarFile jar-file ^JarEntry pom-properties]
+  (with-open [in (.getInputStream jar-file pom-properties)]
+    (->> (doto (Properties.) (.load in))
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:36]: I think we can omit the `->>`:

```clojure
(into {} (doto (Properties.) (.load in)))
```

File: src/leiningen/ring/uberwar.clj [hunk 29]
```diff
@@ -13,26 +17,48 @@
...
+         (into {}))))
+
+(defn- jar-name-from-pom-properties
+  [^JarFile jar-file ^JarEntry pom-properties]
+  (let [{:strs [groupId artifactId version]} (read-jar-pom-properties
+                                               jar-file
+                                               pom-properties)]
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:43]: Probably a little more concise to write:

```clojure
(let [{:strs [groupId artifactId version]}
      (read-jar-pom-properties jar-file pom-properties)]
```

File: src/leiningen/ring/uberwar.clj [hunk 30]
```diff
@@ -13,26 +17,48 @@
...
+    (when (and groupId artifactId version)
+      (str groupId \- artifactId \- version ".jar"))))
+
+(defn- war-path-for-jar [^File jar]
+  (with-open [jar-file (JarFile. jar)]
+    ;; Servlet container will have it's own servlet-api impl
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:49]:
```
;; Servlet container will have its own servlet-api implementation
```

File: src/leiningen/ring/uberwar.clj [hunk 31]
```diff
@@ -13,26 +17,48 @@
...
+    (when-not (contains-javax-servlet-class? jar-file)
+      (let [jar-entries (enumeration-seq (.entries jar-file))
+            name-from-pom (some->> (some pom-properties? jar-entries)
+                                   (jar-name-from-pom-properties jar-file))]
+        (->> (or name-from-pom (.getName jar))
+             (str "WEB-INF/lib/"))))))
+
+(defn- populate-war-with-dependent-jars
+  [war project]
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:58]: Arguments can be put on previous line.

File: src/leiningen/ring/uberwar.clj [hunk 32]
```diff
@@ -42,4 +68,5 @@
      (war/war
        project war-name
        :profiles-to-merge [:uberjar]
-       :additional-writes jar-entries)))
+       :additional-writes populate-war-with-dependent-jars)))
+
```
Comment (weavejester) [inline src/leiningen/ring/uberwar.clj:72]: Additional newline can be omitted.

File: src/leiningen/ring/util.clj [hunk 33]
```diff
@@ -4,7 +4,8 @@
             [clojure.java.io :as io]
             [leiningen.deps]
             [leiningen.core.main :as lein-core]
-            [leinjacker.utils :as lju]))
+            [leinjacker.utils :as lju])
+  (:import (java.io File)))
```
Comment (weavejester) [inline src/leiningen/ring/util.clj:8]: For consistency, `[java.io File]`.

File: src/leiningen/ring/util.clj [hunk 34]
```diff
@@ -31,11 +32,12 @@
           "  :ring {:handler your.app/handler}"))
     (lein-core/exit 1)))
 
-(defn source-file [project namespace]
+(defn source-file
+  ^File [project namespace]
```
Comment (weavejester) [inline src/leiningen/ring/util.clj:36]: No need to put arguments on new line.

**Review by weavejester — CHANGES_REQUESTED** [review]
Looks good - just some minor changes needed for readability/consistency.

**Comment by dpiliouras** [comment]
Latest round of feedback has been incorporated

File: src/leiningen/ring/uberwar.clj (hunk 27 above)
Comment (dpiliouras) [reply src/leiningen/ring/uberwar.clj:27]: `(first (filter ...))` is kind of an anti-pattern... that's why we have `some` right?

**Comment by weavejester** [comment]
Thanks! It looks all ready to merge. Can you squash the commits down into two: one commit that adds type hints to the existing codebase, and the other commit which incorporates the new changes.

**Comment by dpiliouras** [comment]
I'm having some issues rebasing  the already pushed commits. However, as the repo manager I believe you can squash all the commits in a pull request into a single commit by selecting the `Squash and merge` option. Besides, the separation you would like (reflection VS jar-name fixes) would not  clearly reflected in two commits (i.e. my very first commit had a bit of both), so we might as well go for one, with a message like: 

> Reworks `jar-entries` such that jar-name collisions are avoided when possible. Also, adds type-hints to strategic places in order to avoid reflection.

**Comment by weavejester** [comment]
> However, as the repo manager I believe you can squash all the commits in a pull request into a single commit by selecting the `Squash and merge` option.

Unfortunately the "squash and merge" option loses the merge commit, making it largely useless for my purposes. I'd like to keep the merge information (as it's useful), while also having a clean and readable history.

Not everyone is comfortable enough with git to do that, so if you like I can fix the commits and merge locally. Basically I'll just squash everything, then perform a soft reset to unstage everything, and then use `git add -p` to separate the diff out into two commits, in order to ensure that the commits each represent a single change.

GitHub generally isn't smart enough to work out manually merges, so I'll close this commit when I'm done.

**Comment by dpiliouras** [comment]
I'm generally ok with `git`, but on this occasion I came very close to messing up my branch for good. If you can safely do what you describe, that would be fantastic. Many thanks in advance...

**Comment by weavejester** [comment]
Manually merged via e19f19ed
